src/ringmesh/io/geomodel_adapter_resqml.cpp

```diff
@@ -0,0 +1,676 @@
+#include <ringmesh/geomodel/core/geomodel_mesh_entity.h>
+#include <ringmesh/geomodel/core/stratigraphic_column.h>
+
+#include <ringmesh/io/geomodel_adapter_resqml.h>
+
+#include <ringmesh/mesh/mesh_index.h>
+
+#include <ringmesh/mesh/surface_mesh.h>
+#include <ringmesh/mesh/volume_mesh.h>
+
+#include <ringmesh/mesh/mesh_builder.h>
+
+#include <fesapi/common/EpcDocument.h>
+#include <fesapi/common/HdfProxy.h>
+#include <fesapi/resqml2/AbstractFeature.h>
+#include <fesapi/resqml2_0_1/BoundaryFeature.h>
+#include <fesapi/resqml2_0_1/ContinuousProperty.h>
+#include <fesapi/resqml2_0_1/DiscreteProperty.h>
+#include <fesapi/resqml2_0_1/FrontierFeature.h>
+#include <fesapi/resqml2_0_1/Horizon.h>
+#include <fesapi/resqml2_0_1/HorizonInterpretation.h>
+#include <fesapi/resqml2_0_1/LocalDepth3dCrs.h>
+#include <fesapi/resqml2_0_1/LocalTime3dCrs.h>
+#include <fesapi/resqml2_0_1/StratigraphicColumn.h>
+#include <fesapi/resqml2_0_1/StratigraphicColumnRankInterpretation.h>
+#include <fesapi/resqml2_0_1/StratigraphicUnitFeature.h>
+#include <fesapi/resqml2_0_1/StratigraphicUnitInterpretation.h>
+#include <fesapi/resqml2_0_1/TectonicBoundaryFeature.h>
+#include <fesapi/resqml2_0_1/TriangulatedSetRepresentation.h>
+#include <fesapi/resqml2_0_1/UnstructuredGridRepresentation.h>
+
+#include <fesapi/tools/GuidTools.h>
+
+/*!
+ * @brief Implementation of the class to write GeoModel to
+ * RESQML2 .epc file
+ * @author Wan-Chiu Li
+ */
+
+namespace RINGMesh
+{
+    using namespace RESQML2_0_1_NS;
+    using namespace RESQML2_NS;
+    using namespace COMMON_NS;
+    using GMGE = GeoModelGeologicalEntity< 3 >;
+
+    class GeoModelAdapterRESQMLImpl
+    {
+    public:
+        GeoModelAdapterRESQMLImpl(
+            const GeoModel3D& geomodel, const std::string& filename );
+        ~GeoModelAdapterRESQMLImpl() = default;
+
+        bool init();
+        bool save_file();
+        void serialize();
+        bool write_surfaces();
+        bool write_volumes();
+
+        bool write_property( bool first_patch,
+            std::vector< AbstractValuesProperty* >& properties,
+            GEO::AttributesManager& attri_manager,
+            gsoap_resqml2_0_1::resqml2__IndexableElements attachement,
+            AbstractRepresentation& rep );
+        bool write_stratigraphic_column();
+
+        AbstractFeature* find_or_create_feature( const GMGE& entity );
+        AbstractFeatureInterpretation* find_or_create_interpretation(
+            AbstractFeature& feature );
+
+    private:
+        const GeoModel3D& geomodel_;
+        const std::string& filename_;
+        std::unique_ptr< EpcDocument > pck_;
+        AbstractHdfProxy* hdf_proxy_;
+        LocalDepth3dCrs* local_3d_crs_;
+
+        std::map< gmge_id, AbstractFeature* > geo_entity_2_feature_;
+        std::map< AbstractFeature*, AbstractFeatureInterpretation* >
+            feature_2_interp_;
+        std::map< std::string, PropertyKind* > type_2_property_kind_;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused "type_2_property_kind_" "private" field. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS1068)